File: backend/src/main/java/com/spidosoft/erp/validation/ColumnLimits.java
package com.spidosoft.erp.validation;

/**
 * Length and precision limits taken from the column types shown in the Supplied Requirement Material
 * (docs/04-database-schema.md, evidence E1–E6). These are <b>type limits</b>, not business rules.
 *
 * <p>{@code varchar(max)} columns get no length limit here: the engine is TBD, so its maximum is unknown.
 * {@code decimal(p,s)} maps to {@code @Digits(integer = p - s, fraction = s)}.
 */
public final class ColumnLimits {

    // dbo.ItemMaster (E1, E2)
    public static final int ITEM_CODE_LENGTH = 255;      // ItemCode varchar(255)
    public static final int ITEM_UOM_LENGTH = 255;       // UOM varchar(255)
    public static final int ITEM_HSNCODE_LENGTH = 255;   // HSNCODE varchar(255)
    public static final int ITEM_RAW_MATERIAL_LENGTH = 50; // RawMaterial varchar(50)
    public static final int ITEM_DRAWING_NO_LENGTH = 50; // DrawingNo varchar(50)
    public static final int ITEM_SPECIFICATION_LENGTH = 500; // Specification varchar(500)
    /** GSTRate, PurchaseCost, SellingPrice: decimal(20,2). */
    public static final int DECIMAL_20_2_INTEGER = 18;
    public static final int DECIMAL_20_2_FRACTION = 2;

    // dbo.CustomerMaster (E3)
    public static final int CUSTOMER_PIN_CODE_LENGTH = 20;  // PinCode varchar(20)
    public static final int CUSTOMER_TELEPHONE_LENGTH = 20; // Telephone varchar(20)
    public static final int CUSTOMER_MOBILE_LENGTH = 20;    // Mobile varchar(20)

    // dbo.SupplierMaster (E5) — numeric types differ from CustomerMaster (docs/10 C7, not reconciled)
    public static final int SUPPLIER_PIN_CODE_DIGITS = 6;   // PinCode numeric(6,0)
    public static final int SUPPLIER_TELEPHONE_DIGITS = 18; // Telephone numeric(18,0)
    public static final int SUPPLIER_MOBILE_DIGITS = 10;    // Mobile numeric(10,0)

    private ColumnLimits() {
    }
}

File: frontend/src/utils/columnLimits.ts
/**
 * Confirmed column type limits (docs/04, evidence E1–E6). Mirrors backend
 * com.spidosoft.erp.validation.ColumnLimits. These are type limits, NOT business rules.
 * varchar(max) columns have no limit here (engine TBD).
 */
export const ColumnLimits = {
  item: {
    ItemCode: 255,
    UOM: 255,
    HSNCODE: 255,
    RawMaterial: 50,
    DrawingNo: 50,
    Specification: 500,
  },
  /** decimal(20,2): 18 integer digits, 2 fraction digits. */
  decimal20_2: { integer: 18, fraction: 2 },
  customer: { PinCode: 20, Telephone: 20, Mobile: 20 },
  /** numeric(p,0) digit counts. */
  supplier: { PinCode: 6, Telephone: 18, Mobile: 10 },
} as const
